// visitor-js-server/lib/target-adapter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var payloadToTargetKeyMap = {
    MCMID: "marketingCloudVisitorId",
    MCAAMB: "mboxAAMB",
    MCAAMLH: "mboxMCGLH",
    sdid: "mboxMCSDID",
    puuid: "mboxAAMDPUUID",
    pid: "mboxAAMDPID",
    authStateId: "vst.authStateId",
    authState: "vst.authState.authState"
};

function createTargetSchema() {
    return {
        "marketingCloudVisitorId": undefined,
        "mboxParameters": {
            "mboxAAMB": undefined,
            "mboxMCGLH": undefined,
            "mboxAAMDPUUID": undefined,
            "mboxAAMDPID": undefined,
            "vst.authStateId": undefined,
            "vst.authState.authState": undefined,
            "mboxMCSDID": undefined
        }
    };
}

function renameKeys(obj, keysMap) {
    return Object.assign.apply(Object, [{}].concat(_toConsumableArray(Object.keys(obj).map(function (key) {
        return _defineProperty({}, keysMap[key], obj[key]);
    }))));
}

var isObject = function isObject(val) {
    return val === Object(val);
};
var filterFalsyProps = function filterFalsyProps(obj) {
    return JSON.parse(JSON.stringify(obj));
};

function populate(schema, objWithValues) {
    Object.keys(schema).forEach(function (key) {
        if (isObject(schema[key])) {
            // Value is an object, populate it's properties.
            populate(schema[key], objWithValues);
        } else if (objWithValues.hasOwnProperty(key)) {
            schema[key] = objWithValues[key];
        }
    });

    return filterFalsyProps(schema);
}

// Target adapter will output an object in the format of `targetSchema`.
function adapt(obj) {
    var objWithTargetKeys = renameKeys(obj, payloadToTargetKeyMap);
    var targetSchema = createTargetSchema();
    return populate(targetSchema, objWithTargetKeys);
}

exports.default = { adapt: adapt };
module.exports = exports["default"];
