// visitor-js-server/lib/Visitor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = Visitor;

var _visitorJsShared = require("visitor-js-shared");

var _visitorJsShared2 = _interopRequireDefault(_visitorJsShared);

var _cookies = require("./cookies");

var _visitorPayload = require("./visitor-payload");

var _visitorPayload2 = _interopRequireDefault(_visitorPayload);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function Visitor() {
    var _this = this;

    var orgID = arguments.length <= 0 || arguments[0] === undefined ? required("Visitor requires an Org ID!") : arguments[0];

    orgID = normalizeOrgId(orgID);
    var state = {};

    this.generateSupplementalID = function () {
        var sdid = _visitorJsShared2.default.generateSupplementalID();
        Object.assign(state, { sdid: sdid });
        return sdid;
    };

    this.generateVisitorPayload = function (amcvCookie) {
        var sdid = _this.generateSupplementalID();
        var amcvValues = (0, _cookies.parseValuesFromAmcvCookie)(amcvCookie, keysForAmcvValues);
        var customerIDs = state.customerIDs;

        return (0, _visitorPayload2.default)({ sdid: sdid, amcvValues: amcvValues, customerIDs: customerIDs });
    };

    this.getState = function () {
        return _defineProperty({}, orgID, state);
    };

    this.setCustomerIDs = function () {
        var customerIDs = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

        // Store customer IDs as originally provided in state to be shared with client later on.
        Object.assign(state, { customerIDs: customerIDs });
    };
}

var keysForAmcvValues = ["MCMID", "MCAAMB", "MCAAMLH"];

var normalizeOrgId = function normalizeOrgId(orgID) {
    return orgID.indexOf("@") < 0 ? orgID + "@AdobeOrg" : orgID;
};

var required = function required(msg) {
    throw new Error(msg);
};
module.exports = exports["default"];
