// @adobe-mcid/visitor-js-server/lib/Visitor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _visitorJsShared = require("visitor-js-shared");

var _amcvCookie = require("./amcv-cookie");

var _visitorPayload = require("./visitor-payload");

var _visitorPayload2 = _interopRequireDefault(_visitorPayload);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function Visitor() {
    var orgID = arguments.length <= 0 || arguments[0] === undefined ? required("Visitor requires an Org ID!") : arguments[0];

    orgID = normalizeOrgId(orgID);
    var state = {};

    var generateSDID = function generateSDID() {
        var consumerID = arguments.length <= 0 || arguments[0] === undefined ? required("consumerID is required") : arguments[0];

        consumerID = "" + CONSUMER_ID_BASE + consumerID;
        var sdid = (0, _visitorJsShared.generateSupplementalID)();
        Object.assign(state, { sdid: sdid, consumerID: consumerID });
        return sdid;
    };

    this.generatePayload = function (consumerID, amcvCookie) {
        var sdid = generateSDID(consumerID);
        var amcvValues = (0, _amcvCookie.pickValues)(amcvCookie, keysForAmcvValues);
        var customerIDs = state.customerIDs;

        return (0, _visitorPayload2.default)({ sdid: sdid, amcvValues: amcvValues, customerIDs: customerIDs });
    };

    this.getState = function () {
        return _defineProperty({}, orgID, state);
    };

    this.setCustomerIDs = function () {
        var customerIDs = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

        // Store customer IDs as originally provided in state to be shared with client later on.
        Object.assign(state, { customerIDs: customerIDs });
    };
}

Visitor.AuthState = _visitorJsShared.AuthState;

exports.default = Visitor;


var CONSUMER_ID_BASE = "payload:";

var keysForAmcvValues = ["MCMID", "MCAAMB", "MCAAMLH"];

var normalizeOrgId = function normalizeOrgId(orgID) {
    return orgID.indexOf("@") < 0 ? orgID + "@AdobeOrg" : orgID;
};

var required = function required(msg) {
    throw new Error(msg);
};
module.exports = exports["default"];
